### PR TITLE
feat: [PL-32075]: retry background migration

### PR DESCRIPTION
**Pull request by bhavya181**
## Describe your changes
retry background migration, changed the lock to infinite lock with periodic refresh of 30 seconds and wait time of 180 minutes, so, that each pod will wait for 180 minutes while trying to get lock for background migration and if one pod which has lock dies within 180 minutes, another existing pod will be able to take the lock and complete background migration.

Tested locally via adding a migration have sleep time of 30min which means that migration will get completed in 30 min, and I closed the container beforre 30 minutes, started the container again, it was able to take lock successfully and complete the migration.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)

### CODE REVIEW

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 153 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsSLs91NM-yi4nbeUf:-NTsSLs91NM-yi4nbeUg:b-l111p7) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L153)):*
> ```Java
>         MigrationType migrationType = migrationDetail.getMigrationTypeName();
>         try (AcquiredLock ignore = persistentLocker.tryToAcquireInfiniteLockWithPeriodicRefresh(
>                  "Background-" + migrationType, ofMinutes(120))) {
> ```

1. Why did we remove 

    "Background-" + NG_SCHEMA_ID + microservice

     Any reason?


2. Also, what if lock is not acquired with-in 120 mins
___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 154 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsTOaE-1DPKGR1qtGl:-NTsTOaE-1DPKGR1qtGm:b1agwak) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L154)):*
> ```Java
>         try (AcquiredLock ignore = persistentLocker.tryToAcquireInfiniteLockWithPeriodicRefresh(
>                  "Background-" + migrationType, ofMinutes(120))) {
>           HTimeLimiter.callInterruptible21(timeLimiter, Duration.ofHours(2), () -> {
> ```

Why is this 2 hours? Should it be time limited?
___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 169 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsTbMP3LHIRVVhvHer:-NTsTbMP3LHIRVVhvHes:b-s4yw5z) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L169)):*
> ```Java
>       Map<Integer, Class<? extends NGMigration>> migrations, MigrationType migrationTypeName,
>       Class<? extends NGSchema> schemaClass, String serviceName) {
>     if (isBackground) {
> ```

This code should be in the caller. This method should be generic

**Review by bhavya181 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 153 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsSLs91NM-yi4nbeUf:-NTsx7Eq7mJ8yRec7XAY:b-ng0c23) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L153)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

1. Why did we remove 

    "Background-" + NG_SCHEMA_ID + microservice

     Any reason?


2. Also, what if lock is not acquired with-in 120 mins


</blockquote></details>

1. done
2. 120 minutes is long time, lock should be taken within that limit, if its not taken, then no migration will run, we will just get a warning
___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 154 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsTOaE-1DPKGR1qtGl:-NTsxOxg4wbwvMBeiOUu:boy2lth) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L154)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Why is this 2 hours? Should it be time limited?
</blockquote></details>

need to check why its time limited call, @meenakshi-harness  can you please help here.
___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 169 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsTbMP3LHIRVVhvHer:-NTsxUnt387t91ZbCk9i:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L169)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

This code should be in the caller. This method should be generic
</blockquote></details>

Done.

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 153 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsSLs91NM-yi4nbeUf:-NTvTUcCET3Hco3Pw_rK:b4x1nk6) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L153)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

1. done
2. 120 minutes is long time, lock should be taken within that limit, if its not taken, then no migration will run, we will just get a warning
</blockquote></details>

Didn't understand. So, if lock is not required with-in 2 hours, we will attempt to acquire lock on this pod again or not?

**Review by bhavya181 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 153 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsSLs91NM-yi4nbeUf:-NTwNFpq6OTqnEYFWu0d:b4kqpf4) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L153)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Didn't understand. So, if lock is not required with-in 2 hours, we will attempt to acquire lock on this pod again or not?
</blockquote></details>

no, it will just wait for 2 hours, if lock is not acquire it will not retry after that

**Review by bhavya181 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 153 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsSLs91NM-yi4nbeUf:-NTwi3dI1CYHKwa1aqQY:bok6r7r) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L153)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

So it not solve our problem completely? If migrations run beyond 2 hours and pod crashes, our original issue will not be solved. What are the other options?
</blockquote></details>

we can maybe increase wait time to 5 or 6 hours

**Review by bhavya181 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 153 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsSLs91NM-yi4nbeUf:-NTwiANV03KZK17fid3D:b-gaeko3) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L153)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

we can maybe increase wait time to 5 or 6 hours
</blockquote></details>

there is no other way to retry it

**Review by bhavya181 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 153 at r1](https://reviewable.io/reviews/harness/harness-core/46947#-NTsSLs91NM-yi4nbeUf:-NTxoDvg3taz2ZMlG0cR:b-bwrx52) ([raw file](https://github.com/harness/harness-core/blob/aaa6b2729f6b3a9c069d068eb9dcd459437e36f3/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L153)):*
<details><summary><i>Previously, bhavya181 (Bhavya Agrawal) wrote…</i></summary><blockquote>

there is no other way to retry it
</blockquote></details>

increased waiting time to 3 hours

**Comment by bhavya181** [comment]
trigger messagecheck

File: 958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java [hunk 1]
```diff
@@ -149,10 +149,10 @@ private void runBackgroundMigrations(int currentVersion, int maxVersion,
     if (currentVersion < maxVersion) {
       executorService.submit(() -> {
         MigrationType migrationType = migrationDetail.getMigrationTypeName();
-        try (AcquiredLock ignore = persistentLocker.acquireLock(
-                 NGSchema.class, "Background-" + NG_SCHEMA_ID + microservice + migrationType, ofMinutes(120 + 1))) {
+        try (AcquiredLock ignore = persistentLocker.tryToAcquireInfiniteLockWithPeriodicRefresh(
+                 "Background-" + NG_SCHEMA_ID + microservice + migrationType, ofMinutes(180))) {
           HTimeLimiter.callInterruptible21(timeLimiter, Duration.ofHours(2), () -> {
```
Comment (meenakshi-harness) [inline 958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java:154]: This is just for the safety to ensure thread is not busy for too long to execute any BG migration. We can increase this as well to 180 if needed.

**Review by meenakshi-harness — APPROVED** [review]

**Comment by meenakshi-harness** [comment]
Acc. to these changes thread in each pod will be busy checking for lock every 30 secs. If where is a way we can increase this that should help. Rest looks good. cc: @ankit-harness

**Review by ankit-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 154 at r3](https://reviewable.io/reviews/harness/harness-core/46947#gh-1179303888:-NU4tIXsC8WOUqQap37x:b-83wetx) ([raw file](https://github.com/harness/harness-core/blob/2d6ed269343fc5ef5208906635849c1725db37c8/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L154)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

This is just for the safety to ensure thread is not busy for too long to execute any BG migration. We can increase this as well to 180 if needed. 
</blockquote></details>

@meenakshi-harness Shall we remove it altogether? Our migrations doesn't support cancellation. So, this has no effect anyways right now. In fact, same migration can get triggered because of this on 2 different pods if that migration run beyond 2 hours.

**Review by bhavya181 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java` line 154 at r3](https://reviewable.io/reviews/harness/harness-core/46947#gh-1179303888:-NU5cw909N8kObC8is8Q:b-5dfryo) ([raw file](https://github.com/harness/harness-core/blob/2d6ed269343fc5ef5208906635849c1725db37c8/958-migration-sdk/src/main/java/io/harness/migration/service/impl/NGMigrationServiceImpl.java#L154)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

@meenakshi-harness Shall we remove it altogether? Our migrations doesn't support cancellation. So, this has no effect anyways right now. In fact, same migration can get triggered because of this on 2 different pods if that migration run beyond 2 hours. 
</blockquote></details>

removed it

**Review by meenakshi-harness — APPROVED** [review]

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947#-:-NU63fFg71yDAThXx06F:bnfp4nl)* status: 1 file(s) to review. 3. unresolved discussions.

**Review by ankit-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46947)* status: 1 file(s) to review. 2. unresolved discussions.

**Pull request merged**